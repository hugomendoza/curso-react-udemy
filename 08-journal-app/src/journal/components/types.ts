export interface PropsNavBar {
  drawerWidth: number
}

export interface notesProps {
  id: number
  title: string
  body: string
}
