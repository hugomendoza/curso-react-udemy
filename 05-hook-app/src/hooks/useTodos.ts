import { useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = () => {
  const storageItems = JSON.parse( localStorage.getItem("todos")!) 
  return storageItems || [];
}

export const useTodos = ( initialState = [] ) => {
  
  const [ todos, dispatch ] = useReducer( todoReducer, initialState, init );

  
  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos])
  
  
  const handleNewTodo = (todo:{}) => {
    const action = {
      type: "[TODO] Add Todo",
      payload: todo
    }

    dispatch(action);
  }

  const handleDeleteTodo = (id:number) => {
    dispatch({
      type: "[TODO] Remove Todo",
      payload: id
    })
  }

  const handleToggleTodo = ( id:number ) => {
    dispatch({
      type: "[TODO] Toggle Todo",
      payload: id
    })
  }

  const todosCount = todos.length
  // const pendingTodosCount = todos.filter( todo => todo.done === false ).length
  const pendingTodosCount = todos.filter( (todo: { done: boolean; }) => !todo.done).length

  return {
    todos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
    todosCount,
    pendingTodosCount
  }
}
