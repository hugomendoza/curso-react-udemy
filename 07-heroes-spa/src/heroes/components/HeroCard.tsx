import { Link } from "react-router-dom";

export interface PropsHeroes {
  id: string;
  superhero: string;
  publisher: string;
  alter_ego: string;
  first_appearance: string;
  characters: string
};

type CharactersHeroProps = Omit<PropsHeroes, "id" | "superhero" | "publisher" | "first_appearance">;

const CharactersByHero = ({alter_ego, characters}:CharactersHeroProps) => {
  return ( alter_ego === characters)
    ? <></>
    : <p>{ characters }</p>
};

export const HeroCard = ({
  id,
  superhero,
  publisher,
  alter_ego,
  first_appearance,
  characters,
}:PropsHeroes) => {

  const heroImageUrl = `/assets/heroes/${id}.jpg`;

  // const charactersByHero = (<p>{ characters }</p>);

  return (
    <div className="col animate__animated animate__fadeIn">
      <div className="card">
        <div className="row no-gutters">
          <div className="col-4">
            <img
              src={heroImageUrl}
              className="card-img"
              alt={superhero}
            />
          </div>
          <div className="col-8">
            <div className="card-body">
              <h5 className="card-title">{ superhero }</h5>
              <p className="card-text">{ alter_ego }</p>
              {/* {
                ( alter_ego !== characters ) && charactersByHero
              } */}
              <CharactersByHero alter_ego={alter_ego} characters={characters} />
              <p className="card-text">
                <small className="text-muted">{ first_appearance }</small>
              </p>
              <Link to={`/hero/${id}`}>
                MÃ¡s...
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

const defaultProps:PropsHeroes = {
  id: "",
  superhero: "",
  publisher: "",
  alter_ego: "",
  first_appearance:"",
  characters:"",
}

HeroCard.defaultProps = defaultProps;
