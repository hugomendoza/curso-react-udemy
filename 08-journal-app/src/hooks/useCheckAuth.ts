import { useEffect } from "react";
import { onAuthStateChanged } from "firebase/auth";
import { FirebaseAuth } from "../firebase/config";
import { login, logout } from "../store/auth";
import { useAppDispatch, useAppSelector } from "./useRedux";
import { startLoadingNotes } from "../store/journal";

export const useCheckAuth = () => {
  const { status } = useAppSelector( state => state.auth );
  const dispatch = useAppDispatch();

  

  useEffect(() => {
    
    onAuthStateChanged( FirebaseAuth, async( user ) => {
      if( !user ) return dispatch( logout);
      const { uid, email, displayName, photoURL } = user;
      dispatch( login({ uid, email, displayName, photoURL }));
      dispatch( startLoadingNotes ());
    })

  }, [])

  return {
    status
  }
}
