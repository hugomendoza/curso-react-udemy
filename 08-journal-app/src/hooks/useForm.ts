import { useEffect, useMemo, useState } from "react";

interface initialObject {
  email: string;
  password: string;
  displayName?: string;
  index?: any;
}

export const useForm = (initialForm: initialObject, formValidations:any = {}) => {

  const [ formState, setFormState ] = useState(initialForm);
  const [ formValidation, setFormValidation ] = useState({});

  useEffect(() => {
    createValidators();
  }, [ formState ])

  useEffect(() => {
    setFormState(initialForm)
  },[initialForm])

  const isFormValid = useMemo (() => {
    
    for (const formValue of Object.keys ( formValidation)) {
      if ( formValidation[formValue] !== null) return false;
    }

    return true;
  }, [ formValidation ] )
  

  const onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target
    setFormState({
      ...formState,
      [name]: value
    })
  }

  const onResetForm = () => {
    setFormState(initialForm)
  }

  const createValidators = () => {
    const formChechedValues:any = {}

    for (const formField of Object.keys( formValidations )) {
      const [ fn, errorMessage ] = formValidations[formField];

      formChechedValues[`${ formField }Valid`] = fn( formState[formField] ) ? null : errorMessage;
    }

    setFormValidation( formChechedValues );
  }

  return {
    ...formState,
    formState, 
    onInputChange,
    onResetForm,

    ...formValidation,
    isFormValid
  }
}
