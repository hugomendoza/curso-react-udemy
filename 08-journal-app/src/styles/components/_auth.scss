.auth__main {
  align-items: center;
  background-color: $primary;
  display: flex;
  justify-content: center;
  height: 100vh;
  margin: 0;
  width: 100vw;
}

.auth__box-container {
  background-color: white;
  box-shadow: 0px 3px $dark-grey;
  border-radius: 2px;
  padding: 20px;
  width: 250px;
}

.auth__title {
  color: darken($color: $primary, $amount: 20);
  margin-bottom: 20px;
}

.auth__input {
  color: $dark-grey;
  border: 0px;
  border-bottom: 1px solid $light-grey;
  font-size: 16px;
  margin-bottom: 10px;
  height: 20px;
  width: 100%;
  transition: border-bottom .3s ease;

  &:focus {
    outline: none;
    border-bottom: 1px solid $primary;
  }

}

.auth__social-networks {
  align-items: center;
  display: flex;
  justify-content: center;
  flex-direction: column;
  padding: 20px 0;
  width: 100%;
}

.auth__alert-error {
  background-color: red;
  border-radius: 5ox;
  color: white;
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
  padding: 5px;
}