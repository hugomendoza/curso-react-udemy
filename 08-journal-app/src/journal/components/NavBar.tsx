import { useAppDispatch } from "../../hooks";
import {AppBar, Toolbar, IconButton, Grid, Typography} from "@mui/material";
import { LogoutOutlined, MenuOutlined} from "@mui/icons-material"

import { PropsNavBar } from "./types";
import { startLogout } from "../../store/auth";

export const NavBar = ({ drawerWidth = 240 }:PropsNavBar) => {

  const dispatch = useAppDispatch();

  const onLogout = () => {
    dispatch( startLogout() );
  }

  return (
    <AppBar
      position="fixed"
      sx={{
        width: {sm: `calc(100% - ${ drawerWidth}px)`},
        ml: {sm: `${ drawerWidth }px`}
      }}
    >
      <Toolbar>
        <IconButton
          color="inherit"
          edge="start"
          sx={{ mr:2, display: {sm: "none"} }}
        >
          <MenuOutlined />
        </IconButton>

        <Grid
          container
          direction="row"
          justifyContent="space-between"
          alignItems="center"
        >

          <Typography
            variant="h6"
            noWrap
            component="div"
          >
            JournalApp
          </Typography>

          <IconButton
            color="error"
            onClick={ onLogout }
          >
            <LogoutOutlined />
          </IconButton>

        </Grid>

      </Toolbar>
    </AppBar>
  )
}
