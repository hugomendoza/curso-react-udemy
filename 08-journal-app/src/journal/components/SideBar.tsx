import { Box, Divider, Drawer, List, Toolbar, Typography } from "@mui/material";

import { PropsNavBar, notesProps } from "./types";
import { useAppSelector } from "../../hooks";
import SideBarItem from "./SideBarItem";



export const SideBar = ({ drawerWidth }:PropsNavBar) => {
  
  const { displayName } = useAppSelector( state => state.auth );
  const { notes } = useAppSelector( state => state.journal );
  console.log("ðŸš€ ~ file: SideBar.tsx:13 ~ SideBar ~ notes", notes)

  return (
    <Box
      component="nav"
      sx={{ width: { sm: drawerWidth }, flexShrink: {sm: 0} }}
    >
      <Drawer
        variant="permanent" // Review docs for more options
        open // Can be open or open=true
        sx={{
          display: {xs: "block"},
          "& .MuiDrawer-paper": {boxSizing: "border-box", width: drawerWidth}
        }}
      >
        <Toolbar>
          <Typography
            variant="h6"
            noWrap
            component="div"
          >
            { displayName }
          </Typography>
        </Toolbar>

        <Divider />

        <List>
          {
            notes.map( (note:notesProps) => (
              <SideBarItem key={ note.id } {...note}/>
            ))
          }
        </List>

      </Drawer>
    </Box>
  )
}
