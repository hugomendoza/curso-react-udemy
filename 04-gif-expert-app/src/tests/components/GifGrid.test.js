import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');


describe('Pruebas en <GifGrid />', () => {

  const category = 'One Punch';

  test('debe de mostrarse correctamente', () => {
    
    useFetchGifs.mockReturnValue({
      data: [],
      load: true
    });

    const wrapper = shallow( <GifGrid category={ category } />);

    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar items cuando se caragn imÃ¡genes useFetchGifs', () => {

    const gifs = [
      {
        id: 'ABC',
        url: 'http://localhost/cualquier/.cosa.jpg',
        title: 'Cualquier cosa'
      },
      {
        id: 'DEF',
        url: 'http://localhost/cualquier/.cosa-2.jpg',
        title: 'Cualquier cosa 2'
      }
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      load: false
    });
    
    const wrapper = shallow( <GifGrid category={ category }/>);

    expect(wrapper).toMatchSnapshot();
    expect( wrapper.find('p').exists() ).toBe(false);
    expect( wrapper.find('GifGridItem').length ).toBe(gifs.length);

  })
  

});