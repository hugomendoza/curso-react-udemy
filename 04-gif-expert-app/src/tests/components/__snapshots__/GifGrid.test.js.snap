// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> debe de mostrar items cuando se caragn im√°genes useFetchGifs 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="http://localhost/cualquier/.cosa.jpg"
    />
    <GifGridItem
      id="DEF"
      key="DEF"
      title="Cualquier cosa 2"
      url="http://localhost/cualquier/.cosa-2.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <div
    className="card-grid"
  />
</Fragment>
`;
